### PR TITLE
cmd/vendor-install.sh: fix interpreter and RPATH on Linux when necessary

### PR DESCRIPTION
**Pull request by maxim-belkin**
I couldn't install and use vendored Ruby on a Linux system with non-default Homebrew prefix that requires Homebrew's Glibc:

* `RPATH` hardcoded in vendored Ruby points to directories in `/home/linuxbrew/.linuxbrew/`
* Interpreter hardcoded in vendored Ruby points to the host Glibc.

On Linux systems with Glibc prior to 2.23, Homebrew installs a new version of Glibc, so we need to point to that one instead.
On Linux systems with non-`/home/linuxbrew/.linuxbrew` prefix, we need to update `RPATH`.

CC @sjackman

-----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
This seems pretty involved to be in here and I'm wondering why we've never had a user report about this. If a fix is required it feels like it should be in Homebrew/homebrew-portable-ruby. Let's wait and see what @sjackman says.

**Comment by maxim-belkin** [comment]
> Let's wait and see what @sjackman says.

👍 

> This seems pretty involved to be in here

I can briefly explain why a fix is needed here:
1. Vendored Ruby (and, perhaps, other vendored things)
    * uses system Glibc
    * requires Glibc 2.13 and older
2. Homebrew supports systems with "old" Glibc: 
    * 2.13 <= Glibc < 2.23: Homebrew installs and uses `glibc` formula automatically.
    * Glibc < 2.13: warns about Glibc being too old 

Systems with Glibc < 2.13 can bootstrap Ruby and keep using it. If they later try to use vendored Ruby, they will fail because of the `"./$VENDOR_VERSION/bin/$VENDOR_NAME" --version` call below

https://github.com/Homebrew/brew/blob/959d2b7fbb8d15084c0f2e5468396cda2d6b24d5/Library/Homebrew/cmd/vendor-install.sh#L174-L177

The interpreter fix allows Homebrew to install and use vendored Ruby on systems with Glibc < 2.13 but it also makes things consistent on systems with 2.13 <= Glibc < 2.23.

I don't think `RPATH` fix is essential for vendored Ruby but having it set to `/home/linuxbrew/.linuxbrew` on Linux systems with other HOMEBREW_PREFIX isn't right -- I don't think it should be set at all if it's not required (we can remove it when we build Ruby or here).

**Comment by MikeMcQuaid** [comment]
> Systems with Glibc < 2.13 can bootstrap Ruby and keep using it. If they later try to use vendored Ruby, they will fail because of the `"./$VENDOR_VERSION/bin/$VENDOR_NAME" --version` call below

Can you elaborate on what this bug is, how broken it was and on what configurations? Sorry, I'm still not sure I understand why no-one has hit this yet (except you).

> I don't think `RPATH` fix is essential for vendored Ruby but having it set to `/home/linuxbrew/.linuxbrew` on Linux systems with other HOMEBREW_PREFIX isn't right -- I don't think it should be set at all if it's not required (we can remove it when we build Ruby or here).

What breaks if it's set incorrectly? If "nothing": this is definitely a "let's fix in homebrew/homebrew-portable-ruby and not push out a new release just for this".

**Comment by maxim-belkin** [comment]
> > Systems with Glibc < 2.13 can bootstrap Ruby and keep using it. If they later try to use vendored Ruby, they will fail because of the `"./$VENDOR_VERSION/bin/$VENDOR_NAME" --version` call below
> 
> Can you elaborate on what this bug is, how broken it was and on what configurations? Sorry, I'm still not sure I understand why no-one has hit this yet (except you).

I hit this when I tried to switch to using vendored Ruby (from the bootstrapped one) on a system with Glibc older than 2.13 -- vendored Ruby requires Glibc 2.13 and, unfortunately, there is no way to intervene into the installation process because `vendor-install` deletes poured Ruby if the `--version` test fails. The fix seems rather minimal (code-wise) and, at the same time, allows systems with Glibc < 2.13 to use vendored Ruby.

> > I don't think `RPATH` fix is essential for vendored Ruby but having it set to `/home/linuxbrew/.linuxbrew` on Linux systems with other HOMEBREW_PREFIX isn't right -- I don't think it should be set at all if it's not required (we can remove it when we build Ruby or here).
> 
> What breaks if it's set incorrectly? If "nothing": this is definitely a "let's fix in homebrew/homebrew-portable-ruby and not push out a new release just for this".

It depends. If directories specified in `RPATH` don't exist -- that's not a problem. If they do -- that's a security risk because `RPATH` is searched for the required libraries first.

Going back to Linux systems with Glibc < 2.23 in general -- they automatically install `glibc` formula, so we can update the interpreter used by vendored Ruby to the one provided by `glibc` (we can make this step optional for systems with Glibc >= 2.13).

**Comment by MikeMcQuaid** [comment]
This seems like a lot of an code for an edge-case never reported by a (non-maintainer) user. I'd sooner improve the failure message/condition in this case?

**Comment by maxim-belkin** [comment]
> This seems like a lot of an code for an edge-case never reported by a (non-maintainer) user. I'd sooner improve the failure message/condition in this case?

We can do that — I think Homebrew should handle such edge cases somehow rather than leaving them... unhandled. But at the same time I think that taking care of the interpreter is a reasonable thing to do in the context of Homebrew on Linux, especially because we're well positioned to do so: 1) we've got `HOMEBREW_DYNAMIC_LINKER` env. variable that we can use instead of what I did in this PR. 2) we install `glibc` on systems that need it. Let me update this PR to change the interpreter only and let's see how "heavy" it looks. I still hope to hear Shaun's opinion on the proposed changes.

File: Library/Homebrew/cmd/vendor-install.sh [hunk 1]
```diff
@@ -174,6 +174,18 @@ install() {
   tar "$tar_args" "$CACHED_LOCATION"
   safe_cd "$VENDOR_DIR/portable-$VENDOR_NAME"
 
+  if [[ -n $HOMEBREW_LINUX ]]; then
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/vendor-install.sh:177]: Can we get a comment above here explaining what's being done and why?

File: Library/Homebrew/cmd/vendor-install.sh [hunk 2]
```diff
@@ -174,6 +174,18 @@ install() {
...
+    if [[ -x "$HOMEBREW_PREFIX"/opt/patchelf/bin/patchelf ]]; then
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/vendor-install.sh:178]: Should be an error if this doesn't exist, right? How is this going to handle when we're using `patchelf.rb` instead?

**Review by MikeMcQuaid — COMMENTED** [review]
The amount of code here looks more manageable.

**Comment by MikeMcQuaid** [comment]
Nudge on @sjackman for review here when you get a chance.

**Comment by maxim-belkin** [comment]
I did a new install of Homebrew on a system with old Glibc and have to correct myself: Homebrew doesn't auto-install Glibc until the first `brew install` command. Because a user can use `vendor-install` as the first command, I think that `vendor-install` ideally should:
1. check the version of system's Glibc
2. install and use Homebrew's Glibc on a system with Glibc < 2.23.

It would be nice to get @sjackman's opinion on ☝️.

**Comment by sjackman** [comment]
> Nudge on @sjackman for review here when you get a chance.

Thanks for your patience. I got _way_ behind in my GitHub notifications. Feel free to ping me on Slack if you want a review on something that falls in my particular wheelhouse.

**Comment by sjackman** [comment]
I haven't reviewed this PR in detail… but I'll first make a few general comments on the design of `portable-ruby`, and we can discuss more later.

`portable-ruby` uses the host's `glibc` be design. Vendored Ruby is the very first binary package to be installed before anything else, so we cannot rely on any brewed packages being installed. We do not use the brewed `glibc` even after it's installed for consistency, and there's no need. `portable-ruby` uses the host `glibc`.

`portable-ruby` is built using the so-called holy build box approach using the Docker image `debian/eol:wheezy` Debian 7 (Wheezy) and `glibc` 2.13 https://github.com/Linuxbrew/brew/wiki/Bottles#glibc-version. Glibc version 2.13 or later is needed to run `portable-ruby`.

`portable-ruby` uses the host `ld.so` located at `/lib64/ld-linux-x86-64.so.2`

`portable-ruby` should probably not have `RPATH` set to anything, but it doesn't surprise if it were set to `/home/linuxbrew/.linuxbrew/lib` or similar. A PR to fix it if so at https://github.com/Homebrew/homebrew-portable-ruby would be welcome.

**Comment by maxim-belkin** [comment]
> Vendored Ruby is the very first binary package to be installed before anything else

I'm not 100% sure this is still the case, tbh (one has to bootstrap reasonably up-to-date Ruby in order to install Homebrew), but anyway, we need to decide what to do with `vendor-ruby` command on systems with Glibc older than 2.13, so that neither us nor users waste their time. So, our options are:

1. Prevent `vendor-install` from running with an error message `Vendored tools require system Glibc 2.13 or later.`
2. Make `portable-ruby` use Homebrewed `glibc` (as proposed in this PR + make it depend on the version of system glibc + apply necessary fixes in Homebrew/homebrew-portable-ruby).

The first option is easier to implement, but the second one would let users on systems with old Glibc switch to portable Ruby that Homebrew can rely upon. I also think that it might seem strange to users that Homebrew can't use its own packages for certain purposes but this is another issue.

And finally, systems with old Glibc likely have old system Ruby (if at all in some cases) and the fact that  Homebrew runs means that the user bootstrapped Ruby and uses it (_via_ `HOMEBREW_DEVELOPER` and `HOMEBREW_RUBY_PATH`), so the time when `vendor-install` command is invoked doesn't seem to be super relevant.

**Comment by sjackman** [comment]
> I'm not 100% sure this is still the case, tbh (one has to bootstrap reasonably up-to-date Ruby in order to install Homebrew),

I don't believe that's true. The installer is a shell script, and the code to install the vendored Ruby is a shell script, so Ruby is not required to install Homebrew on Linux.

> but anyway, we need to decide what to do with vendor-ruby command on systems with Glibc older than 2.13, so that neither us nor users waste their time. So, our options are:

> Prevent vendor-install from running with an error message Vendored tools require system Glibc 2.13 or later.

This one please. Homebrew does not support systems with host glibc less than 2.13.

https://docs.brew.sh/Homebrew-on-Linux#requirements

> Make portable-ruby use Homebrewed glibc (as proposed in this PR + make it depend on the version of system glibc + apply necessary fixes in Homebrew/homebrew-portable-ruby).

No thank you.

> The first option is easier to implement, but the second one would let users on systems with old Glibc switch to portable Ruby that Homebrew can rely upon. I also think that it might seem strange to users that Homebrew can't use its own packages for certain purposes but this is another issue.

> And finally, systems with old Glibc likely have old system Ruby (if at all in some cases) and the fact that Homebrew runs means that the user bootstrapped Ruby and uses it (via HOMEBREW_DEVELOPER and HOMEBREW_RUBY_PATH), so the time when vendor-install command is invoked doesn't seem to be super relevant.

I'm not sure that I follow this comment. The system Ruby is irrelevant unless it's at least 2.6.3, in which case it can be used.

**Comment by maxim-belkin** [comment]
> I don't believe that's true. The installer is a shell script, and the code to install the vendored Ruby is a shell script, so Ruby is not required to install Homebrew on Linux.

```sh
$ /usr/bin/ruby --version
ruby 1.8.7 (2011-12-28 patchlevel 357) [x86_64-linux]

$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
Homebrew requires Ruby 2.6 which was not found on your system.
Homebrew portable Ruby requires Glibc version 2.13 or newer,
and your Glibc version is too old.
See https://docs.brew.sh/Homebrew-on-Linux#requirements
Install Ruby 2.6 and add its location to your PATH.

```

**Comment by sjackman** [comment]
```
$ /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
Homebrew requires Ruby 2.6 which was not found on your system.
Homebrew portable Ruby requires Glibc version 2.13 or newer,
and your Glibc version is too old.
See https://docs.brew.sh/Homebrew-on-Linux#requirements
Install Ruby 2.6 and add its location to your PATH.
```

I'm glad to see that works as intended. Installing Homebrew on systems with host Glibc older than 2.13 is not supported.

**Comment by MikeMcQuaid** [comment]
> > Prevent vendor-install from running with an error message Vendored tools require system Glibc 2.13 or later.
> 
> This one please. Homebrew does not support systems with host glibc less than 2.13.

Agreed 👍🏻

**Comment by sjackman** [comment]
Superseded by PR https://github.com/Homebrew/brew/pull/9545 `cmd/vendor-install.sh: fail on systems with old Glibc `

**Pull request closed without merging**